### ISSUE TITLE
Implement a hook for changing the stdout/stderr of an interactive command

### ISSUE THREAD

**Issue by andychu** [oilshell/oil#903]
I think this goes in `core/main_loop.py`

Might be related to the DEBUG trap #619, which is a hook for simple commands.

Discussed here:

https://oilshell.zulipchat.com/#narrow/stream/266977-shell-gui/topic/web_shell.20prototype

https://github.com/subhav/web_shell

**Comment by subhav** [comment]
I wonder if we have the same definition of "hook". Just to be clear, I don't think this needs to be a language feature (and I'm not sure what you'd add that can't already be done with redirections on an eval command or on a command group).

It just needs to be a part of the library and whatever future (interprocess?) API you decide to expose to external clients.

**Comment by andychu** [comment]
That's true, but I think hooks are more flexible since they allow you to insert your own code.  Basically you can use a redirect inside this hook, or you could do something else entirely, maybe something like:

```
redirect-stdout() {
  exec >/dev/fd/$GUI_FD
}

trap redirect-stdout BEFORE_INTERACTIVE  # new trap/hook that's like DEBUG, etc.
```

We might also need a way of "sourcing" this on the command line, like:

```
osh --source myhooks.sh --server
```

So `--server` would start the shell in "IPC mode", where it can response to completion requests, query state, etc.  (I believe that IPC is better than APIs, at least for now)

Does that make sense?

I am following the style of bash which is to be more "general".  The things people are doing with PROMPT_COMMAND and the DEBUG trap are pretty unusual.  There are dozens of dozens of programs build on top of these little hooks.

**Comment by andychu** [comment]
Hm I was thinking about this more, and realized that `trap` is probably unnecessary.

If Oil works like a command server, then the GUI could just pass a command right beforehand?  Like if the user types:

```
cd /
ls
```

Then the GUI could do something like:

```
prepare-command 0
cd /
prepare-command 1
ls
```

?   That seems simple enough.  I want to prototype something like this from a CLI but not sure when I'll have time.  I could reuse some of the code from "demoish.py"

https://github.com/oilshell/blog-code/tree/master/interactive-shell

That was a pure Python experiment into the shell UI, which was later folded into Oil itself.  However that code is hard to translate into C++ because of the amount of I/O it does and library integration.  So I could see something like this:

- pure Oil in C++.  Add features to run as a server.
- demoish.py driving it for completion

And then later a GUI could also drive Oil in C++.

I wonder if having the shell interpreter as subprocess of the UI would break anything... hm.

**Comment by subhav** [comment]
> So --server would start the shell in "IPC mode", where it can response to completion requests, query state, etc. (I believe that IPC is better than APIs, at least for now)

So OSH would interpret commands over stdin like usual, but use the other IPC channel to handle things like completion requests out of band? Or, am I misunderstanding?

myhooks.sh in this example sounds like it's basically an OSH command server that's written in OSH. Maybe this is a naïve question, but could you do the same thing with read and eval instead of using debug traps?

**Comment by andychu** [comment]
Hm I haven't really decided on that part yet... It could be stdin, or the server could take commands on the IPC channel too.  A prototype is usually how I figure these issues out.

The `myhooks.sh` example was in case the GUI needs to "prepare" the shell in a different mode.  It could just be a bunch of flags to `osh`, but the equivalent of `source` is a little more flexible.  I think I have wanted that in other circumstances.

Writing an OSH command server in OSH is definitely on the table.  In fact that is https://github.com/oilshell/oil/issues/663  , inspired by ble.sh, which is described a bit here:

https://github.com/oilshell/oil/wiki/How-Interactive-Shells-Work

So all-in-all this is in its early stages, and there are a bunch of other priorities, but I'm looking for people to tell me concretely what is easy to consume from a GUI :)

**Comment by andychu** [comment]
I think headless mode "subsumes" this or at least make it lower priority.  I think the difference is that we were thinking of writing the command server in shell/Oil, which is still valid.  But I thought it was easier to do it with the small `osh --headless` addition
